### PR TITLE
[CI/Build] drop support for  Python 3.8 EOL

### PR DESCRIPTION
**Pull request by aarnphm**
As Python 3.8 reaches [EOL](https://devguide.python.org/versions/), this PR would be a starting point to remove potential code branch on 3.8 only.

I'm still working through all of the code, but this would be a starting point for CI. 

Depends on #8469, and wait til PyTorch drop 3.8 support

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Do we have pyupgrade enabled in the settings for ruff? Would be great if we could use it to get rid of legacy code.

**Comment by aarnphm** [comment]
> Do we have pyupgrade enabled in the settings for ruff? Would be great if we could use it to get rid of legacy code.

I think we can update to latest ruff and use pyupgrade. But I feel like that introduce a lot of merge conflict for other higher priority PR.

For these CI PR would probably be best to run before releasing a new version. What do you think?

ps: I can add ruff tool, but won't run it if that works better.

**Comment by DarkLight1337** [comment]
Currently we use yapf for formatting, and ruff only for linting. I think adding pyupgrade lint rules should not cause many changes to the code.

**Comment by DarkLight1337** [comment]
> I think adding pyupgrade lint rules should not cause many changes to the code.

OK, I stand corrected... maybe we can turn on just a subset of the rules ?

**Comment by DarkLight1337** [comment]
In particular, f-string formatting seems to be responsible for most of the diffs.

**Comment by aarnphm** [comment]
yes, will revert 182d0d7

**Comment by aarnphm** [comment]
fwiw https://docs.astral.sh/ruff/rules/#pyupgrade-up we can exclude f-string format?

**Comment by aarnphm** [comment]
Also are there any specific reason that we locked ruff to 0.1? it is pretty old?

**Comment by DarkLight1337** [comment]
Upgrading `ruff` may add new rules from the categories that we've currently enabled. Let's do that in a separate PR to avoid introducing a huge amount of diffs.

**Comment by DarkLight1337** [comment]
> fwiw https://docs.astral.sh/ruff/rules/#pyupgrade-up we can exclude f-string format?

I think we can add the rules to ignore inside the list in `tool.ruff.lint.ignore`.

**Comment by DarkLight1337** [comment]
Since we might not want to drop Python 3.8 immediately, it's probably better to work on upgrading ruff first.

**Comment by aarnphm** [comment]
#8469 cc @DarkLight1337

**Comment by DarkLight1337** [comment]
Now that Python 3.8 has reached EOL, let's resume work on this.

**Comment by DarkLight1337** [comment]
Heads up that PyTorch 2.5.0 has officially dropped Python 3.8 as well

**Comment by aarnphm** [comment]
yep I will revisit this this weekend, sorry work and other stuff got caught up

**Comment by DarkLight1337** [comment]
@aarnphm any update? Otherwise, we'll work on this issue in #10038 instead.

**Comment by aarnphm** [comment]
Oh I think I already push on my end

**Comment by DarkLight1337** [comment]
Please fix the linter errors. There are also a couple places where we use `sys.version_info`, so it would be great if you could prune the 3.8-only branches!

File: vllm/multimodal/base.py [hunk 1]
```diff
@@ -1,9 +1,22 @@
...
+    NamedTuple,
+    Optional,
+    Tuple,
+    Type,
+    TypedDict,
+    TypeVar,
+    Union,
+    cast,
+    final,
+)
```
Comment (DarkLight1337) [inline vllm/multimodal/base.py:19]: Looks like the formatter changed slightly. Hope this doesn't conflict with yapf/isort...

Comment (aarnphm) [reply vllm/multimodal/base.py:19]: I ran `bash format.sh` and it seems to pass?

Comment (DarkLight1337) [reply vllm/multimodal/base.py:19]: Then it should be fine.

**Comment by aarnphm** [comment]
cc @DarkLight1337 PTAL

**Comment by DarkLight1337** [comment]
Please take a look at the doc failure.

> There are also a couple places where we use sys.version_info, so it would be great if you could prune the 3.8-only code!

And also this.

**Comment by aarnphm** [comment]
Hmm I already remove all sys.version_info compat for `3.8`

**Comment by DarkLight1337** [comment]
> Hmm I already remove all sys.version_info compat for `3.8`

How about `sys.version_info < 3.9`?

**Comment by aarnphm** [comment]
hmm thats weird I thought I already fixed it, anw

**Review by DarkLight1337 — APPROVED** [review]
Thanks for your hard work!

**Pull request merged**